client.go

```diff
@@ -43,30 +47,111 @@ func (f backendFactory) New(remote *config.Backend) proxy.Proxy {
 	return f.bf(remote)
 }
 
-func (backendFactory) newChannel(path string) (*amqp.Channel, closer, error) {
+// connectionHandler handles the connection to the amqp backend
+// the connection is shared by all the requests to that backend
+type connectionHandler struct {
+	logger       logging.Logger
+	logPrefix    string
+	reconnecting *atomic.Bool
+	conn         connection
+}
+
+// newConnectionHandler returns a configured connectionHandler
+func newConnectionHandler(ctx context.Context, l logging.Logger, logPrefix string) connectionHandler {
+	c := connectionHandler{
+		logger:       l,
+		logPrefix:    logPrefix,
+		reconnecting: new(atomic.Bool),
+		conn:         connection{},
+	}
+	go func() {
+		<-ctx.Done()
+		c.conn.Close()
+	}()
+	return c
+}
+
+// newConnection should only be used via connect() beacause it doesn't Lock
+// to replace the connection
+func (h *connectionHandler) newConnection(path string) error {
+	c := connection{}
 	conn, err := amqp.Dial(path)
 	if err != nil {
-		return nil, nopCloser, err
+		return err
 	}
+	c.conn = conn
 	ch, err := conn.Channel()
 	if err != nil {
-		return nil, nopCloser, err
+		return err
+	}
+	c.ch = ch
+	h.conn = c
+	return nil
+}
+
+// connect tries to connect to the service with retries given the configuration
+// strategy
+// WARNING: this function should not be called before setting the reconnecting flag to true
+// using the `CompareAndSwap(false, true)`. as a lock method.
+func (h *connectionHandler) connect(dns string, maxRetries int, bckoff string) error {
+	// This block is executed in the handlers to avoid launching too many goroutines
+	// that will do nothing because of the atomic.Bool set to true
+	// if !h.reconnecting.CompareAndSwap(false, true) {
+	// 	return fmt.Errorf("already reconnecting")
+	// }
+	var res error
+	defer func() {
+		h.reconnecting.Store(false)
+	}()
+	bo := backoff.GetByName(bckoff)
+	h.logger.Debug(h.logPrefix, "connecting to host:", dns)
+	if maxRetries == 0 {
+		maxRetries = 1
+	}
+	for i := 0; i < maxRetries; i++ {
+		<-time.After(bo(i))
+		res = h.newConnection(dns)
+		if res == nil {
+			return nil
```

Comment:
the (re)connect should be logged, as INFO, showing the attempt. `Connection attempt #%d:  Successfully connected`